build-system/tasks/coverage-map.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const argv = require('minimist')(process.argv.slice(2));
+const fs = require('fs').promises;
+const log = require('fancy-log');
+const puppeteer = require('puppeteer');
+const {dist} = require('./dist');
+const {exec} = require('../common/exec');
+const {startServer, stopServer} = require('./serve');
+
+const coverageJsonName = argv.json || 'out.json';
+const testUrl =
+  argv.url || 'http://localhost:8000/examples/everything.amp.html';
+const outHtml = argv.html || 'out.html';
+
+async function doTest() {
+  log(`Opening browser and navigating to ${testUrl}...`);
+  const browser = await puppeteer.launch();
+  const page = await browser.newPage();
+
+  // Enable JavaScript coverage
+  await page.coverage.startJSCoverage();
+  // Navigate to page
+  await page.goto(testUrl);
+  await page.setViewport({
+    width: 1200,
+    height: 800,
+  });
+  log('Scrolling to the end of the page...');
+  await autoScroll(page);
+  log('Testing completed.');
+  const jsCoverage = await page.coverage.stopJSCoverage();
+  const data = JSON.stringify(jsCoverage);
+  log(`Writing to ${coverageJsonName} in dist/${coverageJsonName}...`);
+  await fs.writeFile(`dist/${coverageJsonName}`, data, () => {});
+  await browser.close();
+}
+
+// Source: https://github.com/chenxiaochun/blog/issues/38s
+async function autoScroll(page) {
+  await page.evaluate(async () => {
+    await new Promise((resolve, opt_) => {
+      let totalHeight = 0;
+      const distance = 100;
+      const timer = setInterval(() => {
+        const {scrollHeight} = document.body;
+        /*REVIEW*/
+        window.scrollBy(0, distance);
```

Comment:
This is apparently a banned function, yet this is absolutely necessary because the test itself needs to perform scrolling.
Please advise.

Reply:
It's banned in runtime code. For build system code, you can use `window./*OK*/ scrollBy()`, as suggested in the error message.